### PR TITLE
*: Add config to avoid logging user data to info log

### PR DESCRIPTION
**Pull request by yiwu-arbug**
### What problem does this PR solve?

Add a new config `security.redact-info-log` which avoid printing user data (mostly user keys) to info log. Currently the config doesn't avoid printing user data altogether, but greatly reduce those logs.

Problem Summary:

### What is changed and how it works?

Manually go over all the slog call site and error type messages and wrap those data keys being log with `log_wrappers::Key`. For protobuf structs which can contain user keys and values add a `log_wrappers::ProtobufValue` types which can also filter the output. The behavior is controlled by the `REDACT_INFO_LOG` atomic flag, which is set on tikv server start. If the config is not set by user, the filtering is default on if encryption is enabled and default off otherwise.

### Related changes

No

### Check List 

Tests 

Manually run with `tiup playground` and a light sysbench workload. Some sample input:

default (same as before):
```
[2020/05/07 07:39:13.727 +08:00] [INFO] [peer.rs:327] ["on split"] [split_keys="key 7480000000000000FF0500000000000000F8"] [peer_id=3] [region_id=2]
[2020/05/07 07:39:13.728 +08:00] [INFO] [pd.rs:523] ["try to batch split region"] [task=batch_split] [region="id: 2 region_epoch { conf_ver: 1 version: 1 } peers { id: 3 store_id: 1 }"] [new_region_ids="[new_region_id: 4 new_peer_ids: 5]"] [region_id=2]
[2020/05/07 07:39:13.734 +08:00] [INFO] [apply.rs:1216] ["execute admin command"] [command="cmd_type: BatchSplit splits { requests { split_key: 7480000000000000FF0500000000000000F8 new_region_id: 4 new_peer_ids: 5 } right_derive: true }"] [index=32] [term=6] [peer_id=3] [region_id=2]
[2020/05/07 07:39:13.734 +08:00] [INFO] [apply.rs:1788] ["split region"] [keys="key 7480000000000000FF0500000000000000F8"] [region="id: 2 region_epoch { conf_ver: 1 version: 1 } peers { id: 3 store_id: 1 }"] [peer_id=3] [region_id=2]
[2020/05/07 07:39:13.750 +08:00] [INFO] [peer.rs:1630] ["notify pd with split"] [split_count=2] [peer_id=3] [region_id=2]
[2020/05/07 07:39:13.750 +08:00] [INFO] [peer.rs:1672] ["insert new region"] [region="id: 4 end_key: 7480000000000000FF0500000000000000F8 region_epoch { conf_ver: 1 version: 2 } peers { id: 5 store_id: 1 }"] [region_id=4]
```

`security.redact-info-log=true` or when encryption is enabled:
```
[2020/05/08 08:48:42.301 +08:00] [INFO] [peer.rs:327] ["on split"] [split_keys="key <key>"] [peer_id=3] [region_id=2]
[2020/05/08 08:48:42.301 +08:00] [INFO] [pd.rs:523] ["try to batch split region"] [task=batch_split] [region="id: 2 start_key: <start_key> end_key: <end_key> region_epoch { conf_ver: 1 version: 1 } peers { id: 3 store_id: 1 }"] [new_region_ids="[new_region_id: 4 new_peer_ids: 5]"] [region_id=2]
[2020/05/08 08:48:42.307 +08:00] [INFO] [apply.rs:1239] ["execute admin command"] [command="cmd_type: BatchSplit split {  split_key: <split_key> } prepare_merge { target {  start_key: <start_key> end_key: <end_key> } } commit_merge { source {  start_key: <start_key> end_key: <end_key> } } splits { requests {  split_key: <split_key> new_region_id: 4 new_peer_ids: 5 } right_derive: true }"] [index=32] [term=6] [peer_id=3] [region_id=2]
[2020/05/08 08:48:42.307 +08:00] [INFO] [apply.rs:1816] ["split region"] [keys="key <key>"] [region="id: 2 start_key: <start_key> end_key: <end_key> region_epoch { conf_ver: 1 version: 1 } peers { id: 3 store_id: 1 }"] [peer_id=3] [region_id=2]
[2020/05/08 08:48:42.323 +08:00] [INFO] [peer.rs:1630] ["notify pd with split"] [split_count=2] [peer_id=3] [region_id=2]
[2020/05/08 08:48:42.323 +08:00] [INFO] [peer.rs:1672] ["insert new region"] [region="id: 4 start_key: <start_key> end_key: <end_key> region_epoch { conf_ver: 1 version: 2 } peers { id: 5 store_id: 1 }"] [region_id=4]
```

### Release note 

* Add `security.redact-info-log` config which can reduce logging user data to info log.

### CODE REVIEW

File: components/log_wrappers/src/lib.rs [hunk 1]
```diff
@@ -86,7 +102,119 @@ impl<'a> slog::Value for Key<'a> {
         key: slog::Key,
         serializer: &mut dyn slog::Serializer,
     ) -> slog::Result {
-        serializer.emit_arguments(key, &format_args!("{}", hex::encode_upper(self.0)))
+        if REDACT_INFO_LOG.load(Ordering::SeqCst) {
```
Comment (overvenus) [inline components/log_wrappers/src/lib.rs:105]: How about `Ordering::Relaxed`? As it actually does not require sequentially consistent.

File: components/tikv_util/src/security.rs [hunk 2]
```diff
@@ -24,6 +24,7 @@ pub struct SecurityConfig {
     #[serde(skip)]
     pub override_ssl_target: String,
     pub cert_allowed_cn: HashSet<String>,
+    pub redact_info_log: Option<bool>,
```
Comment (overvenus) [inline components/tikv_util/src/security.rs:27]: What's the difference between `Some(true)`, `Some(false)` and `None`?

File: src/server/gc_worker/gc_manager.rs [hunk 3]
```diff
@@ -527,12 +526,12 @@ impl<S: GcSafePointProvider, R: RegionInfoProvider> GcManager<S, R> {
         // TODO: Find a better way to handle errors. Maybe we should retry.
         debug!(
             "trying gc"; "region_id" => ctx.get_region_id(), "region_epoch" => ?ctx.region_epoch.as_ref(),
-            "end_key" => next_key.as_ref().map(DisplayValue)
+            "end_key" => ?next_key
```
Comment (overvenus) [inline src/server/gc_worker/gc_manager.rs:529]: Does `next_key` contain user data?

File: components/log_wrappers/src/lib.rs (hunk 1 above)
Comment (yiwu-arbug) [reply components/log_wrappers/src/lib.rs:105]: will update

File: components/tikv_util/src/security.rs (hunk 2 above)
Comment (yiwu-arbug) [reply components/tikv_util/src/security.rs:27]: If the option is not set by user, it defaults to false normally to keep existing behavior, but defaults to true if encryption is on.

File: src/server/gc_worker/gc_manager.rs (hunk 3 above)
Comment (yiwu-arbug) [reply src/server/gc_worker/gc_manager.rs:529]: `next_key` is a `txn_types::Key`, and the filtering is done in the type's `Debug` impl. I agree though the inconsistency with normal `Vec<u8>` is confusing and probably error prone..

**Comment by BusyJay** [comment]
I worry that may make debug harder. How about use a check sum instead of hidden it all?

**Comment by yiwu-arbug** [comment]
> I worry that may make debug harder. How about use a check sum instead of hidden it all?

we can. not sure if there will be perf impact.

**Comment by yiwu-arbug** [comment]
> I worry that may make debug harder. How about use a check sum instead of hidden it all?

@overvenus what do you prefer?

**Comment by Yisaer** [comment]
Using checksum is also a optional in pd for log desensitization in order to help debugging. And we also noticed that in https://github.com/pingcap/tidb/pull/19409 , tidb choose using "?" to replace the origin information which is also a optional for PD due to unifying the behavior as a whole system. FYI @yiwu-arbug

File: components/log_wrappers/src/proto.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        let target = format!("{:?}", ProtobufValue(self.0.get_target()));
+        PbPrint::fmt(&target, "target", &mut s);
+        write!(f, "{}", s)
+    }
+}
+
+impl<'a> fmt::Debug for ProtobufValue<'a, Range> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if REDACT_INFO_LOG.load(Ordering::SeqCst) {
```
Comment (overvenus) [inline components/log_wrappers/src/proto.rs:76]: I think it can be Relaxed.

**Comment by skyzh** [comment]
It seems that there're too many conflict files in this PR. Meanwhile, as discussed in https://github.com/tikv/tikv/issues/8469 , protobuf outputs can be redacted by customizing `PbPrint`. Therefore, I would like to open another PR to solve this issue (by only changing `encode_upper` ones).

**Comment by yiwu-arbug** [comment]
Superseded by #8670. Closing.

**Pull request closed without merging**